Feature/image persistence get endpoint

MirelaGeorgiou opened the pull request:
Endpoint to retrieve images saved as part of MaT processes in S3.

Endpoint mainly developed by Liudvikas.

Duslerke committed:
Created new branch for Get Process Image endpoint.

Added really basic controller that lacks a lot.

Duslerke committed:
Added Boundary Response object for Get Process Image endpoint

As discussed, follows typical pattern of giving back the request and generatedAt time in addition to an actual data being returned.

In this case the data returned is simply a string due to the simplicity of endpoint. This means there will probably be no need for the Domain object to wrap this up.

Duslerke committed:
Added a method to Tests helper to generate the Request boundary object for the Get Process Image endpoint.

Duslerke committed:
Made ProcessImageController work with the Usecase.

Controller now calls the usecase.
Controller now calls the usecase with the correct request parameters - no information is lost.
Controller can now grab usecase response and wrap it up before responding.

Duslerke committed:
Made the ProcessImageController work with GetProcessImageRequest Validator.

Controller now calls the validator.
Controller now returns the result based on the validation result:
400 for when request is invalid
200 for the request is valid

Duslerke committed:
Added 'Unexpected' Error Handling logic to ProcessImageController.

Wrote a test that checks whether the information form an error was put and structure and wrapped up into response the way we expect. Test generates 4 possible errors triangulation.

Added implementation. For now it's only for 500 responses. I'd imagine there will be more specific exceptions to catch in the future, however practice show they typically originate from the gateway, hence it's for later.

Duslerke committed:
Factored out error throwing into ErrorThrowerHelper and improved it.

- Created a new tests helper for throwing errors so that error handling would be easier to test.
- Improved the ErrorThrowerHelper to include errors with not null Inner Exceptions.
- Improved the ErrorThrowerHelper to require randomised string for error message triangulation
- Improved both unexpected error handling tests to make use of this helper.

Reason why I don't have random value generation within the helper is because during tests I typically need to call it twice, and I need it so both calls would give the same error message, which was random. The only way to do that is to randomize the message prior to the call and drop it in as an argument.

Duslerke committed:
Added Logger to the 'Process Image Controller' for Get Endpoint

- Logger logs that the request happened.
- Logger logs that the request didn't pass due to validation failure, and returns a formatted message of the errors returned.

- Testing Logger was a bit tricky due to Logger.LogInformation method (and others) being 'extension methods' - whatever that would mean. The thing is, they can't be mocked, so Moq fails both on the .Setup() as well as .Verify(), so what's left is:

- The way to mock an extension method is not to mock an exte nsion method, you instead mock the real method. In this case .Log() with it's biggest constructor. So what actually ends up happening is, instead of checking if the extension method '.LogInformation()' got called, you instead check if the extenstion method called the actual '.Log(LogLevel.Information, ..., ...,.....)' method.

The logic is - if the extension method is called with said parameters, then the real method will be called by that extension method with its parameters passed in, in addition to other hidden parameters that extension method had.

There's slightly better way to do this as I've seen online, however due to time restraints, I won't be able to test those out due it actually being a spike and requiring unknown amount of time.

'https://ardalis.com/testing-logging-in-aspnet-core'

Duslerke committed:
Merge branch 'development' into feature/Image-persistence-Get-endpoint

Duslerke committed:
Added a Usecase Link between Controller and the Gateway for Get Process Image endpoint.

- Updated the definition of the Gateway in the interface, to accept the GetProcessImageRequest boundary.
- Updated the already implemented Gateway (by Mirela) to accept the parameter and use it. Had to do guessing work on the order of parameters supplied into the 'retrieveImage' endpoint as it wasn't documented and it's too late to ask now. If anything, I'll change it tomorrow, though I'm pretty confident on this ordering as it's the same one how every object so far has been defined.
- Wrote the tests for the usecase. This time the usecase is pretty simplistic, so not many tests. Though one of them,... I decided to start checking 'GeneratedAt' date (ideally it should have been checked all along). The check can not be accurate in itself, whoever it can be done to a degree of accuracy - I made my check accurate to 1 second interval, which is a lot. It could be lowered if needed in the future. I definitelly feel that a check is a must, it's just a shame it has to be bit round about one.
- Added Usecase implementation.
- Added regions in Usecase tests to make everything easier to follow.

Duslerke committed:
Added GetProcessImageRequest validator for Get Process Image endpoint.

Duslerke committed:
Added GetProcessImageRequestValidator to dependency Injection

Duslerke committed:
Fixed the PostProcessImageRequestValidator Test from sometimes failing - Modified the request generator.

The test was that there should be no error when the base64 string of correct format was provided. The base64 string provided was generated by this method here in the tests helper.

The issue was that faker file extension generator was sometimes generating file extensions of shorter length than allowed. Due to the method not being flexible enough, had to change it into random string generator.

Had to use string2 and provide char array of allowed characters due to the 'string' method being hard to use for alphanumerics (would have to include @ symbol and something else, if I specified the range from '\u0031' to '\u005a'. Which is not allowed character for a file extension.

MirelaGeorgiou committed:
changes to acommodate new request format

Duslerke commented on mat-process-api.Tests/V1/UseCase/ProcessImageUseCaseTests.cs:
```diff
@@ -76,5 +78,62 @@ public void given_a_request_when_ProcessImageUseCase_ExecutePost_method_is_calle
+        }
+
+        [Test]
+        public void given_a_request_when_ProcessImageUseCase_ExecuteGet_method_is_called_then_it_calls_the_gateway_with_that_string()
```
Test name no longer makes sense. It should be: 
"...then_it_calls_the_gateway_with_the_key_string_generated_from_that_request()", or:
"...then_it_calls_the_gateway_with_the_output_of_imagePersistingHelper()".

Because the way it is know, it's not clear how this string is related to the request that's in the "given_..." part of the name.

Duslerke commented on mat-process-api.Tests/V1/UseCase/ProcessImageUseCaseTests.cs:
```diff
@@ -76,5 +78,62 @@ public void given_a_request_when_ProcessImageUseCase_ExecutePost_method_is_calle
+        {
+            //assert
+            var request = MatProcessDataHelper.CreateGetProcessImageRequestObject();
+            var imageKey = ImagePersistingHelper.generateImageKey(request.processType, request.imageId, request.processRef, request.fileExtension);
```
I think that this helper method should not be static, because due to it being static, it's impossible to mock. 
To have good test coverage we need a test to see if the usecase calls this helper method, but we can't have this test due to it being impossible to mock.

Duslerke commented on mat-process-api/V1/Boundary/GetProcessImageRequest.cs:
```diff
@@ -0,0 +1,16 @@
+        [Required] public string processRef { get; set; }
+        [Required] public string processType { get; set; }
+        [Required] public string imageId { get; set; }
+        [Required] public string fileExtension { get; set; }
```
What are these extra parameters? Our specification say that we should only need "string processRef"  and "string imageId". Do we really need fileExtension, is there no way we could get it from the file on the S3 itself?

![image](https://user-images.githubusercontent.com/43747286/72818453-6b6dac80-3c63-11ea-91d1-df4f572799d8.png)

Duslerke replied:
Just another though I have... doesn't the requiring of fileExtension as parameter imply that we'll need to store that file extension in the process document?

Can't we get the file extension from the s3 response upon file retrieval?
From "HeadersCollection" child object's "Content" field:
![image](https://user-images.githubusercontent.com/43747286/72828075-d8893e00-3c73-11ea-81cf-0b64f25b5a20.png)
, which is inside our S3 response object:
![image](https://user-images.githubusercontent.com/43747286/72828212-1e460680-3c74-11ea-8e6c-63cb2bafd23b.png)

MirelaGeorgiou replied:
We are storing it - when an image get inserted, the key used is a combination of process type, process ref, image ID. The file extension gets appended from the ContentType header. However, when you are retrieving an image, you need to have the complete key (incl file extension) in order to retrieve an image. When we receive a request from the front end, we will need to know what file we are trying to retrieve. This is why file extension is added to the get process image request boundary.

Duslerke commented on mat-process-api/V1/Boundary/GetProcessImageRequest.cs:
```diff
@@ -0,0 +1,16 @@
+    {
+        [Required] public string processRef { get; set; }
+        [Required] public string processType { get; set; }
+        [Required] public string imageId { get; set; }
```
I don't think we need these [Required] attributes. We already have a validator for this exact object that's doing the NotEmpty, NotNull checks. (GetProcessImageRequestValidator)

Duslerke commented on mat-process-api.Tests/V1/Validators/GetProcessImageRequestValidatorTests.cs:
```diff
@@ -0,0 +1,109 @@
+        }
+
+        #region Field is required
+
```
Field is required region is missing 2 tests (that ensure that fields are required) for 2 newly added properties: "processType" and "fileExtension". (For GetProcessImageRequest boundary object)

Duslerke commented on mat-process-api/V1/Controllers/ProcessImageController.cs:
```diff
@@ -53,5 +58,39 @@ public IActionResult PostProcessImage([FromBody] PostProcessImageRequest imageDa
+        [Route("{processType}/{processRef}/{imageId}/{fileExtension}")]
+        [Produces("application/json")]
+        [ProducesResponseType(typeof(GetProcessImageResponse), 200)]
+        public IActionResult GetProcessImage(string processType, string processRef, string imageId,string fileExtension)
```
Are you sure we can't use model binding to map from route to object? Like this, except we would do from route rather than query: ["Can I map multiple query string values to a single controller (object) parameter?"](https://stackoverflow.com/questions/42929376/bind-query-parameters-to-a-model-in-asp-net-core)

Looking at the documentation of [FromRoute()] Attribute, it seems like it supports same properties as [FromQuery()]:
[[FromRoute]](https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.mvc.fromrouteattribute?view=aspnetcore-3.1)
[[FromQuery]](https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.mvc.fromqueryattribute?view=aspnetcore-3.1)
Both have the "Name" property that can be set just like in the SO example I linked above.

This way of handling parameters looks pretty clunky to read and handle. Also makes tests setup a lot more visually heavy to read than prefered. I think we should definitely change this if we can.

Duslerke replied:
Already resolved by @LBHMGeorgieva  in a really elegant way.

Duslerke commented on mat-process-api/V1/Controllers/ProcessImageController.cs:
```diff
@@ -53,5 +58,39 @@ public IActionResult PostProcessImage([FromBody] PostProcessImageRequest imageDa
+        public IActionResult GetProcessImage(string processType, string processRef, string imageId,string fileExtension)
+        {
+            _logger.LogInformation($"Get ProcessImage request for process reference: {processRef} and image Id: {imageId}");
+            var requestData = new GetProcessImageRequest { processType = processType, processRef = processRef, imageId = imageId, fileExtension = fileExtension };
```
If we decide not to do model binding, which I'm feeling personally against, then we should at least make a constructor overload for 4 parameters for GetProcessImageRequest object. We'd need a no parameter constructor as well for clearer test setup.

Duslerke replied:
Obsolete due to model binding changes.

Duslerke commented on mat-process-api/V1/Validators/GetProcessImageRequestValidator.cs:
```diff
@@ -0,0 +1,27 @@
+            RuleFor(req => req.processRef).NotNull().WithMessage("Process reference must be provided.").NotEmpty().WithMessage("Process reference must be provided.").Must(ValidateGuid)
+               .WithMessage("You need to provide a valid process reference.");
+            RuleFor(req => req.imageId).NotNull().WithMessage("Image Id must be provided.").NotEmpty().WithMessage("Image Id must be provided.").Must(ValidateGuid)
+                .WithMessage("You need to provide a valid Image Id.");
```
The request validator is lacking 2 Rules for making 2 newly added GetProcessImageRequest properties required. (processType, fileExtension)

MirelaGeorgiou replied:
Actually, the rules will only be required to check that the image ID and process ref are GUIDs. This is because the rest of the variables are now part of the "path" and if you do not supply them, the API throws 404.

Duslerke replied:
Resolved. After a bit of testing it was discovered that it's possible to pass a whitespace (like '%20') into the endpoint call and it would be interpreted by ASP.NET as 'null' value after doing the model binding. It was decided by the team to add the missing validation for 'NotNull' 'NotEmpty' to have more distinct error messages.

Duslerke commented on mat-process-api.Tests/V1/Controllers/ProcessDataImagesControllerTests.cs:
```diff
@@ -175,5 +182,329 @@ public void given_an_exception_thrown_check_that_the_controller_throws_correct_e
+        public void given_a_valid_request_when_GetProcessImage_controller_method_is_called_then_it_returns_a_200_Ok_response()
+        {
+            //arrange
+            var request = MatProcessDataHelper.CreateGetProcessImageRequestObject();
```
This is a nitpicky thing, but it helps a ton, when reading tests. I've seen there were a lot of changes from:
`var request = new GetProcessImageRequest();`, to:
`var request = MatProcessDataHelper.CreateGetProcessImageRequestObject();`

The reason why this line: `var request = new GetProcessImageRequest();` was left as is rather than changed to be generated from helper was that it helps to signify that this test does not depend on what the request object is, and that the only reason the request object is there is to keep the compiler happy so that the test could be run.

Now that it's changed the helper, it's no longer clear if the contents of the request object are of any significance to the test. Also the line of code is much longer now with a lot more happening, which makes it visually heavier to read in the future.

Since this repeats, I'll mark every place that has exact same issue with: '<1>' comment.

Duslerke replied:
Resolved

Duslerke commented on mat-process-api.Tests/V1/Controllers/ProcessDataImagesControllerTests.cs:
```diff
@@ -175,5 +182,329 @@ public void given_an_exception_thrown_check_that_the_controller_throws_correct_e
+                    return new GetProcessImageResponse(null, DateTime.MinValue, null); //dummy return that will never happen due to error being thrown prior to it. It's only needed to keep compiler happy.
+                });
+
+            var request = MatProcessDataHelper.CreateGetProcessImageRequestObject();
```
'<1>'

Like before, the contents of the request object don't matter. It's a test about catching an exception.

> This is a nitpicky thing, but it helps a ton, when reading tests. I've seen there were a lot of changes from:
> `var request = new GetProcessImageRequest();`, to:
> `var request = MatProcessDataHelper.CreateGetProcessImageRequestObject();`
> 
> The reason why this line: `var request = new GetProcessImageRequest();` was left as is rather than changed to be generated from helper was that it helps to signify that this test does not depend on what the request object is, and that the only reason the request object is there is to keep the compiler happy so that the test could be run.
> 
> Now that it's changed the helper, it's no longer clear if the contents of the request object are of any significance to the test. Also the line of code is much longer now with a lot more happening, which makes it visually heavier to read in the future.

Duslerke replied:
Resolved

Duslerke commented on mat-process-api.Tests/V1/Controllers/ProcessDataImagesControllerTests.cs:
```diff
@@ -175,5 +182,329 @@ public void given_an_exception_thrown_check_that_the_controller_throws_correct_e
+            _mockGetValidator.Setup(x => x.Validate(It.IsAny<GetProcessImageRequest>())).Returns(new FV.ValidationResult()); //setup validator to return a no error validation result
+
+            //act
+            var controllerResponse = _processImageController.GetProcessImage(request.processType,request.processRef,request.imageId,request.fileExtension);
```
Related to comment '<1>' . For this test the contents of the request don't matter, I think this controller method call should be supplied with nulls to indicate that to avoid confusion about the test. Also it makes test harder to read and follow.

Like that:
`var controllerResponse = _processImageController.GetProcessImage(null, null, null ,null);`

Duslerke replied:
Obsolete due to model binding changes.

Duslerke commented on mat-process-api.Tests/V1/Controllers/ProcessDataImagesControllerTests.cs:
```diff
@@ -175,5 +182,329 @@ public void given_an_exception_thrown_check_that_the_controller_throws_correct_e
+            _mockGetValidator.Setup(x => x.Validate(It.IsAny<GetProcessImageRequest>())).Returns(new FV.ValidationResult()); //validation successful
+
+            //act
+            var controllerResponse = _processImageController.GetProcessImage(request.processType,request.processRef,request.imageId, request.fileExtension);
```
**This is repeated comment, but same issue in different place:**

> Related to comment '<1>' . For this test the contents of the request don't matter, I think this controller method call should be supplied with nulls to indicate that to avoid confusion about the test. Also it makes test harder to read and follow.
> 
> Like that:
> `var controllerResponse = _processImageController.GetProcessImage(null, null, null ,null);`

Duslerke replied:
Obsolete due to model binding changes.

MirelaGeorgiou committed:
changed to [FromRoute]

Duslerke commented on mat-process-api.Tests/V1/Controllers/ProcessDataImagesControllerTests.cs:
```diff
@@ -175,5 +182,329 @@ public void given_an_exception_thrown_check_that_the_controller_throws_correct_e
+                    return new GetProcessImageResponse(null, DateTime.MinValue, null); //dummy return that will never happen due to error being thrown prior to it. It's only needed to keep compiler happy.
+                });
+
+            var request = MatProcessDataHelper.CreateGetProcessImageRequestObject();
```
'<1>'

Like before, the contents of the request object don't matter. It's a test about catching an exception messages.

> This is a nitpicky thing, but it helps a ton, when reading tests. I've seen there were a lot of changes from:
> `var request = new GetProcessImageRequest();`, to:
> `var request = MatProcessDataHelper.CreateGetProcessImageRequestObject();`
> 
> The reason why this line: `var request = new GetProcessImageRequest();` was left as is rather than changed to be generated from helper was that it helps to signify that this test does not depend on what the request object is, and that the only reason the request object is there is to keep the compiler happy so that the test could be run.
> 
> Now that it's changed the helper, it's no longer clear if the contents of the request object are of any significance to the test. Also the line of code is much longer now with a lot more happening, which makes it visually heavier to read in the future.

Duslerke commented on mat-process-api.Tests/V1/Controllers/ProcessDataImagesControllerTests.cs:
```diff
@@ -175,5 +182,329 @@ public void given_an_exception_thrown_check_that_the_controller_throws_correct_e
+            _mockGetValidator.Setup(x => x.Validate(It.IsAny<GetProcessImageRequest>())).Returns(new FV.ValidationResult()); //validation successful
+
+            //act
+            var controllerResponse = _processImageController.GetProcessImage(request.processType,request.processRef,request.imageId, request.fileExtension);
```
**This is repeated comment, but same issue in different place:**

> Related to comment '<1>' . For this test the contents of the request don't matter, I think this controller method call should be supplied with nulls to indicate that to avoid confusion about the test. Also it makes test harder to read and follow.
> 
> Like that:
> `var controllerResponse = _processImageController.GetProcessImage(null, null, null ,null);`

Duslerke replied:
Obsolete due to model binding changes.

Duslerke commented on mat-process-api.Tests/V1/Controllers/ProcessDataImagesControllerTests.cs:
```diff
@@ -175,5 +182,329 @@ public void given_an_exception_thrown_check_that_the_controller_throws_correct_e
+            _mockGetValidator.Setup(x => x.Validate(It.IsAny<GetProcessImageRequest>())).Returns(new FV.ValidationResult()); //setup validator to return a no error validation result
+
+            //act
+            _processImageController.GetProcessImage(request.processType,request.processRef,request.imageId, request.fileExtension);
```
**This is repeated comment, but same issue in different place:**

> Related to comment '<1>' . For this test the contents of the request don't matter, I think this controller method call should be supplied with nulls to indicate that to avoid confusion about the test. Also it makes test harder to read and follow.
> 
> Like that:
> `var controllerResponse = _processImageController.GetProcessImage(null, null, null ,null);`

Duslerke replied:
Obsolete due to model binding changes.

Duslerke commented on mat-process-api.Tests/V1/Controllers/ProcessDataImagesControllerTests.cs:
```diff
@@ -175,5 +182,329 @@ public void given_an_exception_thrown_check_that_the_controller_throws_correct_e
+        public void given_a_valid_request_when_GetProcessImage_controller_method_is_called_then_it_calls_usecase()
+        {
+            //arrange
+            GetProcessImageRequest request = MatProcessDataHelper.CreateGetProcessImageRequestObject();
```
'<1>'

Like before, the contents of the request object don't matter. 
It's a test about calling the usecase if validator permits it.

> This is a nitpicky thing, but it helps a ton, when reading tests. I've seen there were a lot of changes from:
> `var request = new GetProcessImageRequest();`, to:
> `var request = MatProcessDataHelper.CreateGetProcessImageRequestObject();`
> 
> The reason why this line: `var request = new GetProcessImageRequest();` was left as is rather than changed to be generated from helper was that it helps to signify that this test does not depend on what the request object is, and that the only reason the request object is there is to keep the compiler happy so that the test could be run.
> 
> Now that it's changed the helper, it's no longer clear if the contents of the request object are of any significance to the test. Also the line of code is much longer now with a lot more happening, which makes it visually heavier to read in the future.

Duslerke commented on mat-process-api.Tests/V1/Controllers/ProcessDataImagesControllerTests.cs:
```diff
@@ -175,5 +182,329 @@ public void given_an_exception_thrown_check_that_the_controller_throws_correct_e
+            _mockGetValidator.Verify(v => v.Validate(It.Is<GetProcessImageRequest>(obj => obj.fileExtension == request.fileExtension)), Times.Once);
+            _mockGetValidator.Verify(v => v.Validate(It.Is<GetProcessImageRequest>(obj => obj.imageId == request.imageId)), Times.Once);
+            _mockGetValidator.Verify(v => v.Validate(It.Is<GetProcessImageRequest>(obj => obj.processRef == request.processRef)), Times.Once);
+            _mockGetValidator.Verify(v => v.Validate(It.Is<GetProcessImageRequest>(obj => obj.processType == request.processType)), Times.Once);
```
You can use **&&** operator from within `It.Is<...>(x => condition1 && condition2 && ... )` expression to validate all these things in a single statement.

Duslerke replied:
Obsolete due to model binding changes.

Duslerke commented on mat-process-api.Tests/V1/Controllers/ProcessDataImagesControllerTests.cs:
```diff
@@ -175,5 +182,329 @@ public void given_an_exception_thrown_check_that_the_controller_throws_correct_e
+        public void when_GetProcessImage_controller_method_is_called_then_it_calls_the_logger()
+        {
+            //arrange
+            var request = MatProcessDataHelper.CreateGetProcessImageRequestObject();
```
'<1>'

Like before, the contents of the request object don't matter. 
It's a test about calling the validator unconditionally at least once. (like when any request comes in - even if it's empty)

> This is a nitpicky thing, but it helps a ton, when reading tests. I've seen there were a lot of changes from:
> `var request = new GetProcessImageRequest();`, to:
> `var request = MatProcessDataHelper.CreateGetProcessImageRequestObject();`
> 
> The reason why this line: `var request = new GetProcessImageRequest();` was left as is rather than changed to be generated from helper was that it helps to signify that this test does not depend on what the request object is, and that the only reason the request object is there is to keep the compiler happy so that the test could be run.
> 
> Now that it's changed the helper, it's no longer clear if the contents of the request object are of any significance to the test. Also the line of code is much longer now with a lot more happening, which makes it visually heavier to read in the future.

Duslerke commented on mat-process-api.Tests/V1/Controllers/ProcessDataImagesControllerTests.cs:
```diff
@@ -175,5 +182,329 @@ public void given_an_exception_thrown_check_that_the_controller_throws_correct_e
+            _mockGetValidator.Setup(l => l.Validate(It.IsAny<GetProcessImageRequest>())).Returns(new FV.ValidationResult()); //setup validator to return a no error validation result
+
+            //act
+            _processImageController.GetProcessImage(request.processType,request.processRef,request.imageId, request.fileExtension);
```
**This is repeated comment, but same issue in different place:**

> Related to comment '<1>' . For this test the contents of the request don't matter, I think this controller method call should be supplied with nulls to indicate that to avoid confusion about the test. Also it makes test harder to read and follow.
> 
> Like that:
> `var controllerResponse = _processImageController.GetProcessImage(null, null, null ,null);`

Duslerke replied:
Obsolete due to model binding changes.

Duslerke reviewed (commented):
Cool! The model binding is done much more elegantly than I thought it could be 👍

Duslerke committed:
Improved tests readability by explicitly stating that test does not depend on parameters of request object

The reason why this line:

`var request = new GetProcessImageRequest();`

was left as is rather than changed to be generated from helper was that it helps to signify that this test does not depend on what the request object is, and that the only reason the request object is there is to keep the compiler happy so that the test could be run.

Clarification: even though some of these tests changed here have `given_a_valid/invalid_request_object_...` in their name, it does not mean that the test itself depends on that object. In most these cases change here, test actually depends on the mock of some dependancy that would would give some sort of response under the conditions of `given_valid/invalid_requets_...`.

The reason why test has that bit in it's name is to justify the mock setup (and what it returns), as that mock under real circumstances should be behaving in a that way when the controller is given valid/invalid response.

If we were to make the tests in a way that the contents of the request object were of some significance, we would fall into the trap of seting up a chain of mocks that have no choice but to have actual implementation within them. Under these circumstances, the test would no longer be testing just that one thing it says it tests, but it would be testing almost the entire controller end to end, which is not what we want from unit tests.

Reasoning behind these particular changes:
- Ln192: "Test only depends the validator mock setup"
- Ln304: "Test about catching an exception, which we throw manually"
- Ln341: "Test about catching exception messages that we generate from exceptions that we manually throw."
- Ln360: "Test about calling the usecase if validator permits it - test depends only on validator mock setup."
- Ln406: "Test about calling the logger unconditionally at least once. (like when any request comes in - even if it's empty)"

Writing this all out made me think that I should improve the test names to better reflect what they're testing.

MirelaGeorgiou committed:
changes as per PR comments

MirelaGeorgiou committed:
Merge branch 'feature/Image-persistence-Get-endpoint' of https://github.com/LBHackney-IT/mat-process-api into feature/Image-persistence-Get-endpoint

Duslerke reviewed (approved):
It's a working endpoint. Further minor changes will be included in the future PR due time restrictions.
